### PR TITLE
Blocks: Add blocks "slash" autocomplete on default block

### PR DESCRIPTION
**Pull request by aduth**
Related: #834
Supersedes: #2284

This pull request seeks to enable a user to add a block by typing a slash `/` in a new default paragraph block, followed by a search parameter. Using arrow keys and click/enter on a block result will replace the paragraph block with a new block of the selected type.

autocomplete

__Implementation notes:__

There's a bit of reinventing the wheel here, but unfortunately I wasn't able to reuse very much:

- I wanted to use [Downshift](https://github.com/paypal/downshift) to manage the typeahead behavior, but it does not support `contenteditable`, only `input`
- I wanted to use a utility for retrieving caret position, but [textarea-caret-position](https://github.com/component/textarea-caret-position) applies only to textarea and input, not `contenteditable`, and [Caret.js](https://github.com/ichord/Caret.js) is dependent on jQuery
- I wanted to use [DropdownMenu](https://github.com/WordPress/gutenberg/tree/master/components/dropdown-menu) for the arrow navigation of options, but I needed only menu behaviors, not the toggling button
- I wanted to use [Popover](https://github.com/WordPress/gutenberg/tree/master/components/popover) to manage positioning, but the position of the autocomplete menu cannot be determined relative a specific node (i.e. retrieving position of text caret requires custom logic)

There may still be opportunity for refactoring to consolidate some of these behaviors.

__Testing instructions:__

Verify that you can insert a new block from an empty paragraph block by typing slash, followed by a block search.

1. Navigate to Gutenberg > New Post
2. Click "Write your story"
3. Type `/`
4. Note that the autocomplete shows
5. Type a block search
6. Note that the autocomplete filters
7. Use arrow keys to navigate results
8. Click or press enter to select block
9. Note that the empty paragraph block is replaced with a new block of the selected type

### CODE REVIEW

**Comment by nicolinuxfr** [comment]
From a user point of view, I really like the idea. I think it could be a great way to add blocks. 👍

**Comment by aduth** [comment]
Ideally, I think this should show the new block as selected after making a replacement. In fact, it is selected, but the `isTyping` state doesn't reset, so focus borders are hidden until the user moves their mouse over the block again. This is made difficult because `isTyping` only tracks a simple boolean `true` / `false` state and a `REPLACE_BLOCKS` action can't be tracked back to occurring within the autocomplete typeahead.

**Comment by mtias** [comment]
> In fact, it is selected, but the isTyping state doesn't reset, so focus borders are hidden until the user moves their mouse over the block again.

I think this is fine, it respects the flow of writing.

File: components/autocomplete/index.js [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+		// Blur is applied to the wrapper node, since if the child is Editable,
+		// the event will not have `relatedTarget` assigned.
+		return (
+			<div
+				ref={ this.bindNode }
+				onBlur={ this.onBlur }
+				className={ classes }
+			>
+				{ cloneElement( Children.only( children ), {
```
Comment (youknowriad) [inline components/autocomplete/index.js]: IIRC, the input (or conteneditable) should have the following aria attributes:


```
role="combobox"
aria-expanded={ isExpanded }
aria-autocomplete="list"
aria-owns={ suggestionsListID }
aria-activedescendant={ selectedSuggestionListItemId }
```
the list should have

```
role="listbox"
```

and each list item should have

```
role="option"
aria-selected={ index === this.props.selectedIndex }
```

Could mimic the `FormTokenField` component

**Review by youknowriad — COMMENTED** [review]
Really good job, works great.

I also see a lot of refactor opportunities:

 - I think we can use the `Popover` component and avoid following the cursor position (That's how slack work)

 - the `dropdown` component should probably be splitted, and reused accross different places (FormtTokenField, Autocomplete, Table Block, Switcher)

**Comment by youknowriad** [comment]
Made some updates to reuse the `popover` component. (The autocomplete is now aware of the window boundaries)

I'm postponing the accessibility changes, because this component is a bit "special". my question here is: Should we add all these combobox attributes knowing that this component won't behave as an autocomplete component if we do not type "/". It's just the paragraph block. I was thinking adding these attributes could be confusing. @afercia 

Anyway, I think we can merge this and address the remaining reusability and accessibility issues separately.

**Comment by karmatosed** [comment]
First, really enjoying using this! Great work everyone on this.

I did find one thing that maybe is more of a 'tweak' than a bug. If you do the following:

- New Gutenberg post.
- Type in /head to get the slash inserter.
- Select heading.
- Type in /heading.. you just get the words /heading

I can see a possible user flow of changing blocks through the slash command. Is this expected or a bug?

2017-09-04 at 19 05

**Comment by youknowriad** [comment]
@karmatosed this is expected as, this is only implemented for the paragraph block (the default block). Extending to other blocks is possible (opt-in).

**Comment by karmatosed** [comment]
Ah crumbs, totally missed it was just for default block! That totally explains it. Whilst this works for a v1, I think that adding into all blocks pretty soon after is important. I won't be only one getting hit by an expectation failure.

**Comment by afercia** [comment]
@youknowriad 
> Should we add all these combobox attributes

Just tested and well, I guess you know the answer 🙂  
When typing `/` and start typing, the block interaction changes drastically and nothing of what happens there get exposed to assistive technologies.
Actually, the block changes in a search field with suggestions, exactly like the tags suggestions and should work the same even from an a11y perspective.
It is, actually, a combobox with suggestions and should use all the ARIA already implemented for such interaction.

**Pull request merged**